@here/harp-mapview/lib/geometry/DisplacedMesh.ts

```diff
@@ -0,0 +1,96 @@
+import * as THREE from "three";
+import { DisplacedBufferAttribute } from "./DisplacedBufferAttribute";
+import { DisplacedBufferGeometry, DisplacementRange } from "./DisplacedBufferGeometry";
+
+/**
+ * Mesh with geometry modified by a displacement map. Overrides raycasting behaviour to apply
+ * displacement map before intersection test.
+ * @internal
+ */
+export class DisplacedMesh extends THREE.Mesh {
+    private static displacedPositions?: DisplacedBufferAttribute;
+
+    private static getDisplacedPositionAttribute(
+        geometry: THREE.BufferGeometry,
+        displacementMap: THREE.DataTexture
+    ): DisplacedBufferAttribute {
+        // Reuse same buffer attribute for all meshes since it's only needed during the
+        // intersection test.
+        if (!DisplacedMesh.displacedPositions) {
+            DisplacedMesh.displacedPositions = new DisplacedBufferAttribute(
+                geometry.attributes.position,
+                geometry.attributes.normal,
+                geometry.attributes.uv,
+                displacementMap
+            );
+        } else {
+            DisplacedMesh.displacedPositions.reset(
+                geometry.attributes.position,
+                geometry.attributes.normal,
+                geometry.attributes.uv,
+                displacementMap
+            );
+        }
+        return DisplacedMesh.displacedPositions;
+    }
+
+    m_displacedGeometry?: DisplacedBufferGeometry;
+
+    /**
+     * Creates an instance of displaced mesh.
+     * @param m_getDisplacementRange Displacement values range getter.
+     * @param [geometry] Original geometry to displace.
+     * @param [material] Material(s) to be used by the mesh. All must have the same displacement
+     * map.
+     */
+    constructor(
+        private m_getDisplacementRange: () => DisplacementRange,
+        geometry?: THREE.Geometry | THREE.BufferGeometry,
+        material?: THREE.Material | THREE.Material[]
+    ) {
+        super(geometry, material);
+    }
+
+    /** @override */
+    raycast(raycaster: THREE.Raycaster, intersects: THREE.Intersection[]): void {
+        // All materials in the object are expected to have the same displacement map.
+        const material: THREE.Material = Array.isArray(this.material)
+            ? this.material[0]
+            : this.material;
+
+        // Use default raycasting implementation if some type is unexpected.
+        if (
+            !(this.geometry instanceof THREE.BufferGeometry) ||
+            !hasDisplacementFeature(material) ||
+            !(material.displacementMap instanceof THREE.DataTexture)
+        ) {
+            super.raycast(raycaster, intersects);
+            return;
+        }
+        const displacementMap = material.displacementMap;
+        const displacementRange = this.m_getDisplacementRange();
+
+        if (this.m_displacedGeometry) {
+            this.m_displacedGeometry.reset(this.geometry, displacementMap, displacementRange);
```

Comment:
Why it is needed? Do we expect displacementMap or displacementRange to change? I see that getter functor is const (without interface for change), so only displacementMap remains.. could you motivate it, just to make we are on the same path.

Reply:
Everything can change: displacementMap/Range due to terrain fallbacks and the geometry itself may also change since it's public in THREE.Mesh.